### DISCUSSION TITLE
Ability to do "export" type commands?

### DISCUSSION THREAD

**Discussion by freakytoad1** [PaloAltoNetworks/pango#64] [Q&A]
Trying to generate a tech support file via PanGo. The api path is `/api/?type=export&category=tech-support`. I see PanGo has Op, Get, Edit, Commit, etc commands if you want to pass through actual API commands. I don't see the same for export. Am I missing something or does PanGo not have this functionality yet?

**Comment by shinmog** [comment]
Export isn't there yet, can add it in tho.

**Reply by freakytoad1** [reply]
While investigating this I saw that this request cannot target a firewall through panorama. The request has to be done against the firewall itself. FYI

`https://<panorama-ip>/api/?type=export&target=<serial-num>&category=tech-support&REST_API_TOKEN=<token>`

```
<response status="error" code="400">
    <result>
        <msg>Exporting tech-support from a remote device is not supported.</msg>
    </result>
</response>
```

**Reply by shinmog** [reply]
You just set the [client.Target](https://github.com/PaloAltoNetworks/pango/blob/master/client.go#L73) param.

You should be using a Firewall object since you're effectively running commands against the firewall.  If you're using any of the [connect](https://github.com/PaloAltoNetworks/pango/blob/master/connect.go) functions it will give you a Firewall instance back.

**Reply by freakytoad1** [reply]
My example above wasn't using PanGo since export isn't a supported type. Just wanted to show that exporting a tech support file via panorama targeting wasn't possible. If you are saying that is possible, instead of individually connecting to a firewall's IP, let me know. I would prefer to go through panorama for as many commands as possible.

The connect function you linked, how does that differ from the suggested method of:
```
    var err error

    c := &pango.Firewall{Client: pango.Client{
        Hostname: "127.0.0.1",
        Username: "admin",
        Password: "admin",
        Logging: pango.LogAction | pango.LogOp,
    }}
    if err = c.Initialize(); err != nil {
        log.Printf("Failed to initialize client: %s", err)
        return
    }
    log.Printf("Initialize ok")
```

Is it a way to connect without knowing if the device is a panorama or firewall, or is there something else I'm missing?

**Reply by shinmog** [reply]
Ah, sorry, didn't look at the response, just the request.  So no, seems like it can't be proxied through Panorama then.

As for connecting without knowing what it is, yep you can use Connect that I linked above, just have to type cast the response to use it:

```go
c := pango.Client{
  Hostname: "127.0.0.1",
  Username: "admin",
  Password: "admin",
}

pci, err := pango.Connect(c)
if err != nil {
  log.Fatal(err)
}

switch con := pci.(type) {
case *pango.Firewall:
  // do firewall stuff
case *pango.Panorama:
  // do panorama stuff
}
```

**Comment by shinmog** [comment]
Export support is now present in pango.

**Reply by freakytoad1** [reply]
Awesome, when is the next release of pango expected?

**Reply by shinmog** [reply]
Well, it's been 6 months since the last release, maybe I should just go ahead and do one soon...  So, soon..?

**Reply by shinmog** [reply]
Released!
